build/AfterCommonTargets.targets

```diff
@@ -0,0 +1,14 @@
+<Project>
+  <PropertyGroup> 
+    <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
+    <VSTestTargets Condition="'$(UseIntrinsics)' != 'true' and '$(TargetFramework)' == 'netcoreapp3.0'">ignore.targets</VSTestTargets>
```

Comment:
What does this do?  I'm not familiar.  Also, since you are also overriding `VSTest` - isn't that enough?

Reply:
If we keep this property setting, we should factor out `'$(UseIntrinsics)' != 'true' and '$(TargetFramework)' == 'netcoreapp3.0'` into an MSBuild property so it is only declared once.

---
In reply to: [206735143](https://github.com/dotnet/machinelearning/pull/562#discussion_r206735143) [](ancestors = 206735143)

Reply:
It turns out that the place where we try to blank `VSTest` (https://github.com/dotnet/machinelearning/pull/562/files/c6f5f8569ea31bc30e0b7ae54870b9cbf51e2bdb#diff-77ebf330684fd059d7aecec3ce7e6218R19) is processed before the program builds the actual `VSTest`, so without this line, even though we would have blanked `VSTest`, the entire `VSTest` would be built all over again.  Because of this order of the build, one way to solve this issue is to ignore the built targets of `VSTest` to ignore the test projects.  Not sure if it is the best way to do it so feel free to suggest any optimization.

Reply:
I helped him with this yesterday and by doing a pp log I found out that `Microsoft.TestPlatform.targets` where imported after the actual `Empty.targets` file even though they are specified to be imported after the common targets, I don't know if that is the way it is designed or if it is a bug. 

So I found out that there was a `VSTestTargets` overridable property declared in: `15.0\Microsoft.Common.targets\ImportAfter\Microsoft.TestPlatform.ImportAfter.targets:17` that points to `Microsoft.TestPlatform.targets` and this `ImportAfter.targets` file will import that project if the specified path exists. So in order to override the `VSTest` target I suggested to just set that property to a non-existent targets file.

Reply:
So if I'm reading this correctly, that means this line in Empty.targets:

``` <Target Name="VSTest"/>```

Isn't doing anything and can be removed, right?

Also - it would probably make sense to move setting this property into `Empty.targets`.  That would resolve my "we should factor this condition out" comment above as well.  `build/AfterCommonTargets.targets` is responsible for deciding whether Empty.targets should be imported. Empty.targets decides how to ensure no targets are run (which includes setting this property).

Reply:
>So if I'm reading this correctly, that means this line in Empty.targets:
>
>`<Target Name="VSTest"/>`
>
>Isn't doing anything and can be removed, right?

It is actually because in run-tests.proj we're calling the VSTest target on every test project: https://github.com/dotnet/machinelearning/blob/master/test/run-tests.proj#L10

And if we don't declare it, it will complain saying that the VSTest target doesn't exist in both test projects targeting netcoreapp3.0.

> Also - it would probably make sense to move setting this property into Empty.targets. That would resolve my "we should factor this condition out" comment above as well. build/AfterCommonTargets.targets is responsible for deciding whether Empty.targets should be imported. Empty.targets decides how to ensure no targets are run (which includes setting this property).

That is right, no sense having the same condition twice, just declare it when `Empty.targets` is imported, which is conditioned already. 

I also thought of just simply excluding both projects from: https://github.com/dotnet/machinelearning/blob/master/test/run-tests.proj#L5

But it makes more sense to me to do everything inside `Empty.targets` since whenever the repo is ready to build and target netcoreapp3.0 we can just remove this files and we're good to go.